src/custom/state/swap/hooks.ts

```diff
@@ -0,0 +1,224 @@
+import useENS from '@src/hooks/useENS'
+import useToggledVersion, { Version } from '@src/hooks/useToggledVersion'
+import { Field } from '@src/state/swap/actions'
+import { tryParseAmount, useSwapState } from '@src/state/swap/hooks'
+import { useUserSlippageTolerance } from '@src/state/user/hooks'
+import { useCurrencyBalances } from '@src/state/wallet/hooks'
+import { isAddress } from '@src/utils'
+import { computeSlippageAdjustedAmounts } from '@src/utils/prices'
+import { Currency, CurrencyAmount, JSBI, Token, TokenAmount, Trade } from '@uniswap/sdk'
+
+export * from '@src/state/swap/hooks'
+
+const BAD_RECIPIENT_ADDRESSES: string[] = [
+  '0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f', // v2 factory
+  '0xf164fC0Ec4E93095b804a4795bBe1e041497b92a', // v2 router 01
+  '0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D' // v2 router 02
+]
+
+/**
+ * Returns true if any of the pairs or tokens in a trade have the given checksummed address
+ * @param trade to check for the given address
+ * @param checksummedAddress address to check in the pairs and tokens
+ */
+function involvesAddress(trade: Trade, checksummedAddress: string): boolean {
+  return (
+    trade.route.path.some(token => token.address === checksummedAddress) ||
+    trade.route.pairs.some(pair => pair.liquidityToken.address === checksummedAddress)
+  )
+}
+
+function tryValueToCurrency(value?: string, currency?: Currency): CurrencyAmount | undefined {
+  if (!value || !currency) {
+    return undefined
+  }
+  try {
+    if (value !== '0') {
+      return currency instanceof Token ? new TokenAmount(currency, value) : CurrencyAmount.ether(value)
+    }
+  } catch (error) {
+    // should fail if the user specifies too many decimal places of precision (or maybe exceed max uint?)
+    console.debug(`Failed to parse input amount: "${value}"`, error)
+  }
+  // necessary for all paths to return a value
+  return undefined
+}
+
+const useTradeWithTip = ({
+  currencyByInput,
+  inputCurrency,
+  outputCurrency,
+  isExactIn,
+  parsedAmount
+}: {
+  currencyByInput?: Currency
+  inputCurrency?: Currency | null
+  outputCurrency?: Currency | null
+  parsedAmount?: CurrencyAmount
+  isExactIn: boolean
+}) => {
+  const tip = getTip()
+  const parsedFeeAmount = tryParseAmount(tip, inputCurrency || undefined)
+
+  // Shows fee amount in OUTPUT token
+  const feeForTradeExactIn = useTradeExactIn(isExactIn ? parsedFeeAmount : undefined, outputCurrency ?? undefined)
+    ?.inputAmount
+  // Shows fee amount in INPUT token
+  const feeForTradeExactOut = useTradeExactOut(outputCurrency ?? undefined, !isExactIn ? parsedFeeAmount : undefined)
+    ?.inputAmount
+
+  const bestTradeExactIn = useTradeExactIn(
+    isExactIn
+      ? feeForTradeExactIn && parsedAmount
+        ? tryValueToCurrency(
+            // PARSED_USER_AMOUNT - FEE_AMOUNT
+            JSBI.subtract(parsedAmount.raw, feeForTradeExactIn.raw).toString(),
+            currencyByInput
+          )
+        : parsedAmount
+      : undefined,
```

Comment:
This is a little hard to read. Can you instead do `if else` outside of `useTradeExactIn`

Reply:
haha blame them im only duplicating and adjusting as needed :stuck_out_tongue: 

But yes ill do that